Coleção de arquivos escolhida: Tech

Primeiro descompactamos o arquivo .zip com as coleções de arquivos

In [82]:
import zipfile
import os

# Especifique o nome do arquivo .zip que você fez upload
nome_arquivo_zip = "bbc-fulltext.zip"

# Diretório onde o arquivo .zip está localizado
diretorio_raiz = "/content/"

# Diretório de destino para a extração
diretorio_destino = "/content/extracao/"

# Criar o diretório de destino, se ele não existir
os.makedirs(diretorio_destino, exist_ok=True)

# Caminho completo para o arquivo .zip
caminho_arquivo_zip = os.path.join(diretorio_raiz, nome_arquivo_zip)

# Extrair o conteúdo do arquivo .zip
with zipfile.ZipFile(caminho_arquivo_zip, 'r') as zip_ref:
    zip_ref.extractall(diretorio_destino)

print("Arquivo descompactado com sucesso!")


Arquivo descompactado com sucesso!


Depois lemos todos os arquivos escolhidos e salvamos seu texto em uma lista

In [83]:
import os

# Diretório onde os arquivos de texto estão localizados
diretorio_arquivos = "/content/extracao/bbc/tech"

# Lista para armazenar o texto de cada arquivo
documentos = []

# Iterar sobre os arquivos na pasta
for nome_arquivo in os.listdir(diretorio_arquivos):
    caminho_completo = os.path.join(diretorio_arquivos, nome_arquivo)

    if os.path.isfile(caminho_completo):
        # Ler o conteúdo do arquivo
        with open(caminho_completo, 'r') as arquivo:
            conteudo = arquivo.read()
            documentos.append(conteudo)  # Adicionar o conteúdo à lista

print("Texto de todos os arquivos foi armazenado na lista.")


Texto de todos os arquivos foi armazenado na lista.


In [84]:
# Imprimir o número de itens na lista
print("Número de itens na lista de documentos:", len(documentos))

Número de itens na lista de documentos: 401


In [85]:
# Imports a serem utilizados posteriormente
import nltk
nltk.download('stopwords')
from collections import Counter

from nltk.corpus import stopwords
from nltk.stem import RSLPStemmer
from nltk.tokenize import sent_tokenize , word_tokenize
from nltk.stem import PorterStemmer

STOP_WORDS = set(stopwords.words('english'))

stemmer = PorterStemmer()
nltk.download('punkt')

import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


Consultas a serem executadas:

- query1: "Best anti-virus"
- query2: "Apple new Iphone"
- query3: "Stolen credit-card number"
- query4: "Internet scammer"
- query5: "Robotic vehicle"
- query6: "Block spam e-mails"

Query 1 : "Best anti-virus" ⛳

In [87]:
query = "Best anti-virus"

**Primeira rodada: sem retirar stopword e sem "stemming"**

In [88]:
# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso, para evitar duplicados
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.3662 - Título do artigo: Virus poses as Christmas e-mail
Similaridade: 0.3146 - Título do artigo: Joke e-mail virus tricks users
Similaridade: 0.2946 - Título do artigo: Microsoft debuts security tools
Similaridade: 0.2607 - Título do artigo: Toxic web links help virus spread
Similaridade: 0.2409 - Título do artigo: Microsoft seeking spyware trojan
Similaridade: 0.2092 - Título do artigo: Cyber crime booms in 2004
Similaridade: 0.1769 - Título do artigo: BBC leads interactive Bafta wins
Similaridade: 0.1665 - Título do artigo: Beckham virus spotted on the net
Similaridade: 0.1604 - Título do artigo: Half-Life 2 sweeps Bafta awards
Similaridade: 0.1584 - Título do artigo: Windows worm travels with Tetris
Similaridade: 0.1450 - Título do artigo: Voters flock to blog awards site
Similaridade: 0.1297 - Título do artigo: Security warning over 'FBI virus'
Similaridade: 0.1169 - Título do artigo: Fast moving phone viruses appear
Similarid

É possível observar que 81 resultados foram retornados, com a maior similaridade de 0.3662 e a menor 0.0088.

**Segunda rodada: retirando stopwords e sem "stemming"**

In [90]:
#Stop words

lista_sem_stop = []
for cada_texto in documentos:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)


# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.3994 - Título do artigo: Virus poses as Christmas e-mail
Similaridade: 0.3315 - Título do artigo: Joke e-mail virus tricks users
Similaridade: 0.3167 - Título do artigo: Microsoft debuts security tools
Similaridade: 0.2832 - Título do artigo: Toxic web links help virus spread
Similaridade: 0.2545 - Título do artigo: Microsoft seeking spyware trojan
Similaridade: 0.2417 - Título do artigo: Cyber crime booms in 2004
Similaridade: 0.1898 - Título do artigo: BBC leads interactive Bafta wins
Similaridade: 0.1832 - Título do artigo: Beckham virus spotted on the net
Similaridade: 0.1787 - Título do artigo: Half-Life 2 sweeps Bafta awards
Similaridade: 0.1702 - Título do artigo: Windows worm travels with Tetris
Similaridade: 0.1592 - Título do artigo: Voters flock to blog awards site
Similaridade: 0.1402 - Título do artigo: Security warning over 'FBI virus'
Similaridade: 0.1252 - Título do artigo: Fast moving phone viruses appear
Similarid

É possível perceber que a similaridade aumentou agora retirando as stop words dos textos. A maior similaridade ficou em 0.3994 e a menor em 0.0095. O número de resultados retornado ficou o mesmo.

**Terceira rodada: retirando stopwords e  com "stemming"**

In [92]:
# Stemming

def preprocess_text(text):
    words = nltk.word_tokenize(text)
    words = [stemmer.stem(word) for word in words]
    return ' '.join(words)

lista_stemming = []
for cada_elemento in documentos:
    text = (preprocess_text(cada_elemento))
    lista_stemming.append(text)

# Aplicar stemming na consulta
query_stemmed = preprocess_text(query)

#Stop words

lista_sem_stop = []
for cada_texto in lista_stemming:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)

# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query_stemmed])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.4109 - Título do artigo: Virus poses as Christmas e-mail
Similaridade: 0.3520 - Título do artigo: Joke e-mail virus tricks users
Similaridade: 0.2990 - Título do artigo: Microsoft debuts security tools
Similaridade: 0.2892 - Título do artigo: Toxic web links help virus spread
Similaridade: 0.2650 - Título do artigo: Microsoft seeking spyware trojan
Similaridade: 0.2490 - Título do artigo: Cyber crime booms in 2004
Similaridade: 0.1902 - Título do artigo: BBC leads interactive Bafta wins
Similaridade: 0.1802 - Título do artigo: Beckham virus spotted on the net
Similaridade: 0.1740 - Título do artigo: Windows worm travels with Tetris
Similaridade: 0.1619 - Título do artigo: Half-Life 2 sweeps Bafta awards
Similaridade: 0.1430 - Título do artigo: Voters flock to blog awards site
Similaridade: 0.1257 - Título do artigo: Fast moving phone viruses appear
Similaridade: 0.1253 - Título do artigo: Fast moving phone bugs appear
Similaridade:

É possível perceber que a similaridade aumentou ainda mais ao realizar o stemming e retirar as stop words dos textos. A maior similaridade ficou em 0.4109 e a menor em 0.0097. O número de resultados ficou o mesmo.

Query 2 : "Apple new Iphone" ⛳

In [93]:
query = "Apple new Iphone"

**Primeira rodada: sem retirar stopword e sem "stemming"**

In [94]:
# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso, para evitar duplicados
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.1792 - Título do artigo: Apple sues to stop product leaks
Similaridade: 0.1549 - Título do artigo: ITunes user sues Apple over iPod
Similaridade: 0.1493 - Título do artigo: Apple attacked over sources row
Similaridade: 0.1429 - Título do artigo: Apple sues 'Tiger' file sharers
Similaridade: 0.1366 - Título do artigo: Apple makes blogs reveal sources
Similaridade: 0.1135 - Título do artigo: Apple Mac mini gets warm welcome
Similaridade: 0.1058 - Título do artigo: Apple unveils low-cost 'Mac mini'
Similaridade: 0.0993 - Título do artigo: Creator of first Apple Mac dies
Similaridade: 0.0767 - Título do artigo: Hacker threat to Apple's iTunes
Similaridade: 0.0607 - Título do artigo: Apple iPod family expands market
Similaridade: 0.0533 - Título do artigo: Rivals of the £400 Apple...
Similaridade: 0.0527 - Título do artigo: Apple laptop is 'greatest gadget'
Similaridade: 0.0440 - Título do artigo: Gadget show heralds MP3 season
Similari

217 resultados foram retornados, com similaridades de 0.1792 a 0.0016.

**Segunda rodada: retirando stopwords e sem "stemming"**

In [54]:
#Stop words

lista_sem_stop = []
for cada_texto in documentos:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)


# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.1924 - Título do artigo: Apple sues to stop product leaks
Similaridade: 0.1676 - Título do artigo: ITunes user sues Apple over iPod
Similaridade: 0.1616 - Título do artigo: Apple attacked over sources row
Similaridade: 0.1552 - Título do artigo: Apple sues 'Tiger' file sharers
Similaridade: 0.1495 - Título do artigo: Apple makes blogs reveal sources
Similaridade: 0.1275 - Título do artigo: Apple Mac mini gets warm welcome
Similaridade: 0.1156 - Título do artigo: Apple unveils low-cost 'Mac mini'
Similaridade: 0.1085 - Título do artigo: Creator of first Apple Mac dies
Similaridade: 0.0813 - Título do artigo: Hacker threat to Apple's iTunes
Similaridade: 0.0724 - Título do artigo: Apple iPod family expands market
Similaridade: 0.0664 - Título do artigo: Apple laptop is 'greatest gadget'
Similaridade: 0.0627 - Título do artigo: Rivals of the £400 Apple...
Similaridade: 0.0472 - Título do artigo: Gadget show heralds MP3 season
Similari

A similaridade dos resultados aumentou, indo de 0.1924 a 0.0018.

**Terceira rodada: retirando stopwords e  com "stemming"**

In [56]:
# Stemming

def preprocess_text(text):
    words = nltk.word_tokenize(text)
    words = [stemmer.stem(word) for word in words]
    return ' '.join(words)

lista_stemming = []
for cada_elemento in documentos:
    text = (preprocess_text(cada_elemento))
    lista_stemming.append(text)

# Aplicar stemming na query
query_stemmed = preprocess_text(query)

#Stop words

lista_sem_stop = []
for cada_texto in lista_stemming:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)

# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query_stemmed])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.1991 - Título do artigo: Apple sues to stop product leaks
Similaridade: 0.1722 - Título do artigo: ITunes user sues Apple over iPod
Similaridade: 0.1694 - Título do artigo: Apple attacked over sources row
Similaridade: 0.1629 - Título do artigo: Apple makes blogs reveal sources
Similaridade: 0.1592 - Título do artigo: Apple sues 'Tiger' file sharers
Similaridade: 0.1299 - Título do artigo: Apple Mac mini gets warm welcome
Similaridade: 0.1156 - Título do artigo: Apple unveils low-cost 'Mac mini'
Similaridade: 0.1102 - Título do artigo: Creator of first Apple Mac dies
Similaridade: 0.0844 - Título do artigo: Hacker threat to Apple's iTunes
Similaridade: 0.0764 - Título do artigo: Apple iPod family expands market
Similaridade: 0.0670 - Título do artigo: Apple laptop is 'greatest gadget'
Similaridade: 0.0637 - Título do artigo: Rivals of the £400 Apple...
Similaridade: 0.0474 - Título do artigo: Gadget show heralds MP3 season
Similari

A similaridade dos resultados aumentou, indo de 0.1991 a 0.0016. Porém é possível observar que a similaridade dos documentos com menor similaridade diminuiu.

Query 3 : "Stolen credit-card number" ⛳

In [57]:
query = "Stolen credit-card number"

**Primeira rodada: sem retirar stopword e sem "stemming"**

In [58]:
# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso, para evitar duplicados
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.1749 - Título do artigo: Dozens held over ID fraud site
Similaridade: 0.1215 - Título do artigo: Who do you think you are?
Similaridade: 0.0871 - Título do artigo: 'Friends fear' with lost mobiles
Similaridade: 0.0795 - Título do artigo: Loyalty cards idea for TV addicts
Similaridade: 0.0604 - Título do artigo: Cyber crime booms in 2004
Similaridade: 0.0548 - Título do artigo: BT program to beat dialler scams
Similaridade: 0.0537 - Título do artigo: California sets fines for spyware
Similaridade: 0.0537 - Título do artigo: Solutions to net security fears
Similaridade: 0.0535 - Título do artigo: Ban hits Half-Life 2 pirates hard
Similaridade: 0.0508 - Título do artigo: US hacker breaks into T-Mobile
Similaridade: 0.0507 - Título do artigo: Cyber criminals step up the pace
Similaridade: 0.0485 - Título do artigo: China 'to overtake US net use'
Similaridade: 0.0476 - Título do artigo: Software watching while you work
Similaridade: 0.0

136 resultados foram retornados, com similaridades de 0.1749 a 0.0037.

**Segunda rodada: retirando stopwords e sem "stemming"**

In [59]:
#Stop words

lista_sem_stop = []
for cada_texto in documentos:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)


# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.1894 - Título do artigo: Dozens held over ID fraud site
Similaridade: 0.1520 - Título do artigo: Who do you think you are?
Similaridade: 0.0960 - Título do artigo: 'Friends fear' with lost mobiles
Similaridade: 0.0830 - Título do artigo: Loyalty cards idea for TV addicts
Similaridade: 0.0698 - Título do artigo: Cyber crime booms in 2004
Similaridade: 0.0636 - Título do artigo: Solutions to net security fears
Similaridade: 0.0596 - Título do artigo: Ban hits Half-Life 2 pirates hard
Similaridade: 0.0591 - Título do artigo: BT program to beat dialler scams
Similaridade: 0.0574 - Título do artigo: California sets fines for spyware
Similaridade: 0.0545 - Título do artigo: China 'to overtake US net use'
Similaridade: 0.0536 - Título do artigo: Software watching while you work
Similaridade: 0.0532 - Título do artigo: Cyber criminals step up the pace
Similaridade: 0.0532 - Título do artigo: US hacker breaks into T-Mobile
Similaridade: 0.0

A similaridade dos resultados aumentou, indo de 0.1894 a 0.0040.

**Terceira rodada: retirando stopwords e  com "stemming"**

In [70]:
# Stemming

def preprocess_text(text):
    words = nltk.word_tokenize(text)
    words = [stemmer.stem(word) for word in words]
    return ' '.join(words)

lista_stemming = []
for cada_elemento in documentos:
    text = (preprocess_text(cada_elemento))
    lista_stemming.append(text)

# Aplicar stemming na query
query_stemmed = preprocess_text(query)

#Stop words

lista_sem_stop = []
for cada_texto in lista_stemming:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)

# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query_stemmed])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.4700 - Título do artigo: Robots learn 'robotiquette' rules
Similaridade: 0.4278 - Título do artigo: Hitachi unveils 'fastest robot'
Similaridade: 0.3901 - Título do artigo: Robotic pods take on car design
Similaridade: 0.3703 - Título do artigo: Humanoid robot learns how to run
Similaridade: 0.1091 - Título do artigo: GTA sequel is criminally good
Similaridade: 0.0897 - Título do artigo: 2D Metal Slug offers retro fun
Similaridade: 0.0754 - Título do artigo: Gadget show heralds MP3 season
Similaridade: 0.0749 - Título do artigo: Gadget show heralds MP3 Christmas
Similaridade: 0.0644 - Título do artigo: The Force is strong in Battlefront
Similaridade: 0.0618 - Título do artigo: Speech takes on search engines
Similaridade: 0.0572 - Título do artigo: Tech helps disabled speed demons
Similaridade: 0.0369 - Título do artigo: Game makers get Xbox 2 sneak peek
Similaridade: 0.0341 - Título do artigo: 'Brainwave' cap controls computer
Simi

A similaridade dos resultados aumentou, indo de 0.2290 a 0.0036. Porém é possível observar que a similaridade dos documentos com menor similaridade diminuiu.

Query 4 : "Internet Scammers" ⛳

In [61]:
query = "Internet Scammers"

**Primeira rodada: sem retirar stopword e sem "stemming"**

In [62]:
# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso, para evitar duplicados
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.1645 - Título do artigo: Domain system scam fear
Similaridade: 0.1636 - Título do artigo: Domain system opens door to scams
Similaridade: 0.0592 - Título do artigo: Net regulation 'still possible'
Similaridade: 0.0505 - Título do artigo: Musicians 'upbeat' about the net
Similaridade: 0.0414 - Título do artigo: The pirates with no profit motive
Similaridade: 0.0389 - Título do artigo: Call for action on internet scam
Similaridade: 0.0381 - Título do artigo: Security scares spark browser fix
Similaridade: 0.0372 - Título do artigo: Microsoft plans 'safer ID' system
Similaridade: 0.0353 - Título do artigo: Blog reading explodes in America
Similaridade: 0.0350 - Título do artigo: Firefox browser takes on Microsoft
Similaridade: 0.0350 - Título do artigo: Go-ahead for new internet names
Similaridade: 0.0339 - Título do artigo: Can Yahoo dominate next decade?
Similaridade: 0.0327 - Título do artigo: Britons fed up with net service
Simila

103 resultados foram retornados, com similaridades de 0.1645 a 0.0020.

**Segunda rodada: retirando stopwords e sem "stemming"**

In [63]:
#Stop words

lista_sem_stop = []
for cada_texto in documentos:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)


# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.1798 - Título do artigo: Domain system scam fear
Similaridade: 0.1790 - Título do artigo: Domain system opens door to scams
Similaridade: 0.0676 - Título do artigo: Net regulation 'still possible'
Similaridade: 0.0555 - Título do artigo: Musicians 'upbeat' about the net
Similaridade: 0.0483 - Título do artigo: Call for action on internet scam
Similaridade: 0.0481 - Título do artigo: The pirates with no profit motive
Similaridade: 0.0417 - Título do artigo: Security scares spark browser fix
Similaridade: 0.0397 - Título do artigo: Firefox browser takes on Microsoft
Similaridade: 0.0394 - Título do artigo: Microsoft plans 'safer ID' system
Similaridade: 0.0387 - Título do artigo: Podcasts mark rise of DIY radio
Similaridade: 0.0379 - Título do artigo: Can Yahoo dominate next decade?
Similaridade: 0.0379 - Título do artigo: Blog reading explodes in America
Similaridade: 0.0371 - Título do artigo: Go-ahead for new internet names
Simila

A similaridade dos resultados aumentou, indo de 0.1798 a 0.0027.

**Terceira rodada: retirando stopwords e  com "stemming"**

In [71]:
# Stemming

def preprocess_text(text):
    words = nltk.word_tokenize(text)
    words = [stemmer.stem(word) for word in words]
    return ' '.join(words)

lista_stemming = []
for cada_elemento in documentos:
    text = (preprocess_text(cada_elemento))
    lista_stemming.append(text)

# Aplicar stemming na query
query_stemmed = preprocess_text(query)

#Stop words

lista_sem_stop = []
for cada_texto in lista_stemming:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)

# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query_stemmed])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.4700 - Título do artigo: Robots learn 'robotiquette' rules
Similaridade: 0.4278 - Título do artigo: Hitachi unveils 'fastest robot'
Similaridade: 0.3901 - Título do artigo: Robotic pods take on car design
Similaridade: 0.3703 - Título do artigo: Humanoid robot learns how to run
Similaridade: 0.1091 - Título do artigo: GTA sequel is criminally good
Similaridade: 0.0897 - Título do artigo: 2D Metal Slug offers retro fun
Similaridade: 0.0754 - Título do artigo: Gadget show heralds MP3 season
Similaridade: 0.0749 - Título do artigo: Gadget show heralds MP3 Christmas
Similaridade: 0.0644 - Título do artigo: The Force is strong in Battlefront
Similaridade: 0.0618 - Título do artigo: Speech takes on search engines
Similaridade: 0.0572 - Título do artigo: Tech helps disabled speed demons
Similaridade: 0.0369 - Título do artigo: Game makers get Xbox 2 sneak peek
Similaridade: 0.0341 - Título do artigo: 'Brainwave' cap controls computer
Simi

É possível observar que a similaridade diminuiu, devido ao processo de stemming.

Query 5 : "Robotic vehicle" ⛳

In [66]:
query = "Robotic vehicle"

**Primeira rodada: sem retirar stopword e sem "stemming"**

In [67]:
# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso, para evitar duplicados
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.1110 - Título do artigo: Robotic pods take on car design
Similaridade: 0.1000 - Título do artigo: GTA sequel is criminally good
Similaridade: 0.0427 - Título do artigo: 2D Metal Slug offers retro fun
Similaridade: 0.0365 - Título do artigo: 'Brainwave' cap controls computer
Similaridade: 0.0297 - Título do artigo: Speech takes on search engines
Similaridade: 0.0293 - Título do artigo: Tech helps disabled speed demons
Similaridade: 0.0287 - Título do artigo: The Force is strong in Battlefront
Total de resultados impressos: 7


7 resultados foram retornados, com similaridades de 0.1110 a 0.0287.

**Segunda rodada: retirando stopwords e sem "stemming"**

In [68]:
#Stop words

lista_sem_stop = []
for cada_texto in documentos:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)


# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.1305 - Título do artigo: Robotic pods take on car design
Similaridade: 0.1203 - Título do artigo: GTA sequel is criminally good
Similaridade: 0.0466 - Título do artigo: 2D Metal Slug offers retro fun
Similaridade: 0.0385 - Título do artigo: 'Brainwave' cap controls computer
Similaridade: 0.0347 - Título do artigo: The Force is strong in Battlefront
Similaridade: 0.0336 - Título do artigo: Speech takes on search engines
Similaridade: 0.0328 - Título do artigo: Tech helps disabled speed demons
Total de resultados impressos: 7


A similaridade dos resultados aumentou, indo de 0.1305 a 0.0328.

**Terceira rodada: retirando stopwords e  com "stemming"**

In [72]:
# Stemming

def preprocess_text(text):
    words = nltk.word_tokenize(text)
    words = [stemmer.stem(word) for word in words]
    return ' '.join(words)

lista_stemming = []
for cada_elemento in documentos:
    text = (preprocess_text(cada_elemento))
    lista_stemming.append(text)

# Aplicar stemming na query
query_stemmed = preprocess_text(query)

#Stop words

lista_sem_stop = []
for cada_texto in lista_stemming:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)

# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query_stemmed])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.4700 - Título do artigo: Robots learn 'robotiquette' rules
Similaridade: 0.4278 - Título do artigo: Hitachi unveils 'fastest robot'
Similaridade: 0.3901 - Título do artigo: Robotic pods take on car design
Similaridade: 0.3703 - Título do artigo: Humanoid robot learns how to run
Similaridade: 0.1091 - Título do artigo: GTA sequel is criminally good
Similaridade: 0.0897 - Título do artigo: 2D Metal Slug offers retro fun
Similaridade: 0.0754 - Título do artigo: Gadget show heralds MP3 season
Similaridade: 0.0749 - Título do artigo: Gadget show heralds MP3 Christmas
Similaridade: 0.0644 - Título do artigo: The Force is strong in Battlefront
Similaridade: 0.0618 - Título do artigo: Speech takes on search engines
Similaridade: 0.0572 - Título do artigo: Tech helps disabled speed demons
Similaridade: 0.0369 - Título do artigo: Game makers get Xbox 2 sneak peek
Similaridade: 0.0341 - Título do artigo: 'Brainwave' cap controls computer
Simi

É possível observar que a similaridade aumentou muito, indo para 0.4700. Isso foi devido ao processo de stemming.

Query 6 : "Block spam e-mails" ⛳

In [95]:
query = "Block spam e-mails"

**Primeira rodada: sem retirar stopword e sem "stemming"**

In [96]:
# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso, para evitar duplicados
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.3769 - Título do artigo: Spam e-mails tempt net shoppers
Similaridade: 0.3262 - Título do artigo: Bad e-mail habits sustains spam
Similaridade: 0.2107 - Título do artigo: Junk e-mails on relentless rise
Similaridade: 0.1714 - Título do artigo: Screensaver tackles spam websites
Similaridade: 0.1707 - Título do artigo: Warnings about junk mail deluge
Similaridade: 0.1485 - Título do artigo: Anti-spam screensaver scrapped
Similaridade: 0.1432 - Título do artigo: US duo in first spam conviction
Similaridade: 0.1338 - Título do artigo: Security warning over 'FBI virus'
Similaridade: 0.1152 - Título do artigo: Freeze on anti-spam campaign
Similaridade: 0.1127 - Título do artigo: Cyber criminals step up the pace
Similaridade: 0.0954 - Título do artigo: Anti-spam laws bite spammer hard
Similaridade: 0.0942 - Título do artigo: Warning over tsunami aid website
Similaridade: 0.0831 - Título do artigo: US state acts to stop 'spammers'
Similari

42 resultados foram retornados, com similaridades de 0.3969 a 0.0127.

**Segunda rodada: retirando stopwords e sem "stemming"**

In [97]:
#Stop words

lista_sem_stop = []
for cada_texto in documentos:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)


# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.4120 - Título do artigo: Spam e-mails tempt net shoppers
Similaridade: 0.3499 - Título do artigo: Bad e-mail habits sustains spam
Similaridade: 0.2267 - Título do artigo: Junk e-mails on relentless rise
Similaridade: 0.1895 - Título do artigo: Screensaver tackles spam websites
Similaridade: 0.1813 - Título do artigo: Warnings about junk mail deluge
Similaridade: 0.1637 - Título do artigo: Anti-spam screensaver scrapped
Similaridade: 0.1509 - Título do artigo: US duo in first spam conviction
Similaridade: 0.1446 - Título do artigo: Security warning over 'FBI virus'
Similaridade: 0.1270 - Título do artigo: Freeze on anti-spam campaign
Similaridade: 0.1184 - Título do artigo: Cyber criminals step up the pace
Similaridade: 0.1101 - Título do artigo: Warning over tsunami aid website
Similaridade: 0.1001 - Título do artigo: Anti-spam laws bite spammer hard
Similaridade: 0.0889 - Título do artigo: US state acts to stop 'spammers'
Similari

A similaridade dos resultados aumentou bastante, indo de 0.4120 a 0.0149.

**Terceira rodada: retirando stopwords e  com "stemming"**

In [98]:
# Stemming

def preprocess_text(text):
    words = nltk.word_tokenize(text)
    words = [stemmer.stem(word) for word in words]
    return ' '.join(words)

lista_stemming = []
for cada_elemento in documentos:
    text = (preprocess_text(cada_elemento))
    lista_stemming.append(text)

# Aplicar stemming na query
query_stemmed = preprocess_text(query)


#Stop words

lista_sem_stop = []
for cada_texto in lista_stemming:
  cada_texto_tokens = nltk.word_tokenize(cada_texto)
  text_sem_stop = []
  for cada_palavra in cada_texto_tokens:
    if cada_palavra not in STOP_WORDS:
        text_sem_stop.append(cada_palavra)
  lista_sem_stop.append(text_sem_stop)

# Junte as palavras em cada lista da lista_sem_stop para criar strings
documentos_sem_stop = [' '.join(texto) for texto in lista_sem_stop]

# Inicializar o TfidfVectorizer
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documentos_sem_stop + [query_stemmed])

# Calcular a similaridade do cosseno entre a consulta e os documentos
cosine_similarities = cosine_similarity(tfidf_matrix[-1], tfidf_matrix[:-1])

# Obter índices ordenados dos documentos por similaridade decrescente
sorted_indices = np.argsort(cosine_similarities[0])[::-1]

numero_resultados = 0
titulos_impressos = set()  # Conjunto para rastrear os títulos dos documentos impressos

# Exibir os documentos ordenados por similaridade
print("Documentos ordenados por similaridade:")
for idx in sorted_indices:
    similarity = cosine_similarities[0][idx]
    if similarity > 0:  # Verificar se a similaridade é maior que zero
        documento = documentos[idx]  # Recupera o documento da lista original
        primeira_linha = documento.split('\n', 1)[0]  # Pega a primeira linha do documento

        # Verificar se o título já foi impresso
        if primeira_linha not in titulos_impressos:
            print(f"Similaridade: {similarity:.4f} - Título do artigo: {primeira_linha}")
            titulos_impressos.add(primeira_linha)  # Adicionar o título ao conjunto
            numero_resultados += 1  # Incrementa a contagem

print(f"Total de resultados impressos: {numero_resultados}")

Documentos ordenados por similaridade:
Similaridade: 0.5031 - Título do artigo: Bad e-mail habits sustains spam
Similaridade: 0.4453 - Título do artigo: Spam e-mails tempt net shoppers
Similaridade: 0.4173 - Título do artigo: Warnings about junk mail deluge
Similaridade: 0.3063 - Título do artigo: Screensaver tackles spam websites
Similaridade: 0.3003 - Título do artigo: Junk e-mails on relentless rise
Similaridade: 0.2523 - Título do artigo: Anti-spam screensaver scrapped
Similaridade: 0.2399 - Título do artigo: Bush website blocked outside US
Similaridade: 0.2257 - Título do artigo: Attack prompts Bush site block
Similaridade: 0.1889 - Título do artigo: Warning over tsunami aid website
Similaridade: 0.1787 - Título do artigo: Cyber criminals step up the pace
Similaridade: 0.1631 - Título do artigo: US state acts to stop 'spammers'
Similaridade: 0.1600 - Título do artigo: Freeze on anti-spam campaign
Similaridade: 0.1571 - Título do artigo: Anti-spam laws bite spammer hard
Similaridad

É possível observar que a similaridade aumentou com stemming, para 0.5031 e 0.0083. O número de resultados retornado aumentou de 42 para 71, por conta do stemming na palavra "emails".

**Considerações finais:**

- Foi possível observar que todas as rodadas retornaram o mesmo número de resultados, pois quando um termo está presente em um documento, mesmo com a remoção de stop words e com o stemming, ele continuará presente. Porém a query 6 retornou mais resultados com o stemming, por conta da palavra "emails".

- Em todas as queries, exceto a query 4 ("Internet scam"), o nível de similaridade sempre aumentou conforme cada rodada. Geralmente houve um grande salto de similaridade da primeira para a segunda rodada, devido à remoção das stop-words.

- Também foi possível observar que a similaridade dos documentos com menor similaridade diminuiu em todas as queries, exceto na primeira query, na etapa 3. Isso porque o peso dos termos baixa com o stemming.

- As consultas que retornaram mais resultados foram as que possuíam termos menos específicos, como a query 2 ("Apple new Iphone") que por causa do "new" retornou 217 resultados. E as queries com termos menos específicos, como a query 5 ("Robotic vehicle") retornou apenas 7 resultados.